File: gdx/src/com/badlogic/gdx/graphics/g3d/utils/TextureBinder.java
package com.badlogic.gdx.graphics.g3d.utils;

import com.badlogic.gdx.graphics.GLTexture;
import com.badlogic.gdx.graphics.Texture;

/**
 * Responsible for binding textures, may implement a strategy to avoid binding a texture unnecessarily. A TextureBinder may
 * decide to which texture unit it binds a texture.
 *
 *  */
public interface TextureBinder {
    /**
     * Prepares the binder for operation, must be matched with a call to {@link #end()}.
     */
    void begin();

    /**
     * Disables all used texture units and unbinds textures. Resets the counts.
     */
    void end();

    /**
     * Binds the texture to an available unit and applies the filters in the descriptor.
     *
     * @param textureDescriptor the {@link TextureDescriptor}
     * @return the unit the texture was bound to
     */
    int bind(TextureDescriptor textureDescriptor);

    /**
     * Binds the texture to an available unit.
     *
     * @param texture the {@link Texture}
     * @return the unit the texture was bound to
     */
    int bind(GLTexture texture);

    /**
     * @return the number of binds actually executed since the last call to {@link #resetCounts()}
     */
    int getBindCount();

    /**
     * @return the number of binds that could be avoided by reuse
     */
    int getReuseCount();

    /**
     * Resets the bind/reuse counts
     */
    void resetCounts();
}

File: gdx/src/com/badlogic/gdx/graphics/g3d/utils/TextureBinder.kt
package com.badlogic.gdx.graphics.g3d.utils

import com.badlogic.gdx.graphics.GLTexture

/**
 * Info : This interface was moved to Kerman game engine.
 *
 * Responsible for binding textures, may implement a strategy to avoid binding a texture unnecessarily. A TextureBinder may
 * decide to which texture unit it binds a texture.
 */
interface TextureBinder {

    /**
     * Prepares the binder for operation, must be matched with a call to [end].
     */
    fun begin()

    /**
     * Disables all used texture units and unbinds textures. Resets the counts.
     */
    fun end()

    /**
     * Binds the texture to an available unit and applies the filters in the descriptor.
     *
     * @param textureDescriptor the [TextureDescriptor]
     * @return the unit the texture was bound to.
     */
    fun bind(textureDescriptor: TextureDescriptor<*>): Int

    /**
     * Binds the texture to an available unit.
     *
     * @param texture the [GLTexture]
     * @return the unit the texture was bound to
     */
    fun bind(texture: GLTexture): Int

    /**
     * @return the number of binds actually executed since the last call to [resetCounts].
     */
    fun getBindCount(): Int

    /**
     * @return the number of binds that could be avoided by reuse.
     */
    fun getReuseCount(): Int

    /**
     * Resets the bind/reuse counts
     */
    fun resetCounts()

}
